### PR TITLE
[Woo POS] Integrate "include_types" backend API changes for POS Products API call

### LINKED ISSUE #14649: [Woo POS] Integrate "include_types" backend API changes for POS Products API call

### PR DESCRIPTION
**Pull request by jaclync**
Closes: #14649

## Description

This pull request includes adding a new feature flag for variable products support, and updating POS product retrieval API request to include variable products for WC versions 9.6+. For older WC versions, only simple products are returned.

### New Feature Flag:
* Added a new case `variableProductsInPointOfSale` to the `FeatureFlag` enum to support variable products in POS. (`Experiments/Experiments/FeatureFlag.swift` [Experiments/Experiments/FeatureFlag.swiftR210-R213](diffhunk://#diff-d42bbc633193f198cb514f56f2b0fe8ac65787a3e40a56c44b24069217788929R210-R213))
* Updated the `DefaultFeatureFlagService` to enable the `variableProductsInPointOfSale` feature flag for local developer and alpha build configurations. (`Experiments/Experiments/DefaultFeatureFlagService.swift` [Experiments/Experiments/DefaultFeatureFlagService.swiftR100-R101](diffhunk://#diff-26c9afc49bb1f1a168517d43d5e36a64ff45adb046469369ef3e4d6800ccc19bR100-R101))

### Product Retrieval Updates:
* Modified the `ProductsRemote` class to include variable products when the feature flag is enabled, and updated the method signature to accept product types. (`Networking/Networking/Remote/ProductsRemote.swift` [Networking/Networking/Remote/ProductsRemote.swiftL200-R213](diffhunk://#diff-dde78446a8994905cf951ddc694e4b3620a81a6dfd9e719f48afec2cd4c3e7b6L200-R213))
* Added a new constant `productTypes` in `ProductsRemote` to handle multiple product types. (`Networking/Networking/Remote/ProductsRemote.swift` [Networking/Networking/Remote/ProductsRemote.swiftR617](diffhunk://#diff-dde78446a8994905cf951ddc694e4b3620a81a6dfd9e719f48afec2cd4c3e7b6R617))

### POS Product Service Updates:
* Updated the `PointOfSaleProductService` to check the feature flag and request the appropriate product types. (`Yosemite/Yosemite/PointOfSale/PointOfSaleProductService.swift` [[1]](diffhunk://#diff-03cc427764fb8714a46696aed239423dfc04e8b8094e074b8e4c3662dc913531R20-R36) [[2]](diffhunk://#diff-03cc427764fb8714a46696aed239423dfc04e8b8094e074b8e4c3662dc913531L41-R47)
* Modified the `HubMenuViewModel` to pass the feature flag status to the `PointOfSaleProductService`. (`WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuViewModel.swift` [WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuViewModel.swiftL100-R101](diffhunk://#diff-a7353a3c9b54e18ce66e192000f034059a624c6824a359575150dc1788697d04L100-R101))

### Test Updates:
* Updated the `ProductsRemoteTests` to reflect changes in the product retrieval method and to ensure the correct product types are loaded based on the feature flag. (`Networking/NetworkingTests/Remote/ProductsRemoteTests.swift` [[1]](diffhunk://#diff-01c6287fe6f095c7bdd598502a7d5a4d5b7c42b9291b20a415e3a242d2691dafL935-R943) [[2]](diffhunk://#diff-01c6287fe6f095c7bdd598502a7d5a4d5b7c42b9291b20a415e3a242d2691dafL952-R964) [[3]](diffhunk://#diff-01c6287fe6f095c7bdd598502a7d5a4d5b7c42b9291b20a415e3a242d2691dafL973-R973) [[4]](diffhunk://#diff-01c6287fe6f095c7bdd598502a7d5a4d5b7c42b9291b20a415e3a242d2691dafL982-R982) [[5]](diffhunk://#diff-01c6287fe6f095c7bdd598502a7d5a4d5b7c42b9291b20a415e3a242d2691dafL991-R991)
* Added new tests in `PointOfSaleProductServiceTests` to verify that the correct query parameters are set based on the feature flag status. (`Yosemite/YosemiteTests/PointOfSale/POSProductProviderTests.swift` [[1]](diffhunk://#diff-1ba1528eb413467eaecc7aa17e37f46dfc3988342bd6c1460f0ba82ac3feceeaL18-R19) [[2]](diffhunk://#diff-1ba1528eb413467eaecc7aa17e37f46dfc3988342bd6c1460f0ba82ac3feceeaR112-R141)

### WC version 9.6+

Prerequisite: have a store with WC version 9.6+, I find JN site with Beta Tester option + `trunk` in core the easiest way. The store should have at least one purchasable simple & variable product.

🗒️ If trying to access the site with WPCOM login using Jetpack connection, JN sites with dev WC version seem to have `woocommerce_is_active: false` in the `me/sites` API response and thus these sites don't show up in the store picker. I hard-coded [this line](https://github.com/woocommerce/woocommerce-ios/blob/f30517c85795ad30e860e12266300ca7ed202b13/Networking/Networking/Model/Site.swift#L108) to be `true` for all sites and also the [POS eligibility](https://github.com/woocommerce/woocommerce-ios/blob/f30517c85795ad30e860e12266300ca7ed202b13/WooCommerce/Classes/ViewRelated/Dashboard/Settings/POS/POSEligibilityChecker.swift#L19) to return `Just(true).eraseToAnyPublisher()` as a workaround locally.

- Switch to the store in the prerequisite
- Go to Menu > Point of Sale mode --> the variable product(s) should appear in the product selector, along with simple product(s). Support for adding variable products to cart & payment is not available yet

### WC version < 9.6

Prerequisite: have a store with WC version < 9.6. The store should have at least one purchasable simple & variable product.

- Switch to the store in the prerequisite
- Go to Menu > Point of Sale mode --> the variable product(s) should **not** appear in the product selector. Only simple products should be listed

## Testing information

- [ ] @jaclync tests when the feature flag is disabled, only simple products are loaded for WC version 9.6+
- [ ] @jaclync tests when the feature flag is disabled, only simple products are loaded for WC version < 9.6

## Screenshots

WC version 9.6+, feature flag disabled | WC version 9.6+, feature flag enabled
-- | --
Simulator Screenshot - iPad Pro 11-inch (M4) - 2024-12-11 at 12 21 05 | Simulator Screenshot - iPad Pro 11-inch (M4) - 2024-12-11 at 12 19 38

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Networking/Networking/Remote/ProductsRemote.swift [hunk 1]
```diff
@@ -197,17 +197,20 @@ public final class ProductsRemote: Remote, ProductsRemoteProtocol {
...
     /// - siteID: Site for which we'll fetch remote products.
+    /// - productTypes: A list of product types to be included in the results.
     /// - pageNumber: Number of page that should be retrieved.
     ///
-    public func loadSimpleProductsForPointOfSale(for siteID: Int64, pageNumber: Int = 1) async throws -> [Product] {
+    public func loadProductsForPointOfSale(for siteID: Int64, productTypes: [ProductType] = [.simple], pageNumber: Int = 1) async throws -> [Product] {
         let parameters = [
             ParameterKey.page: String(pageNumber),
             ParameterKey.perPage: POSConstants.productsPerPage,
+            // When both productType and productTypes are provided, the productType is ignored in WC versions 9.6+.
```
Comment (staskus) [inline Networking/Networking/Remote/ProductsRemote.swift:211]: I missed the fact that we got the functionality of fetching multiple specific types into 9.6, great to see 👍

**Review by staskus — APPROVED** [review]
🙇 Thanks for an excellent description with explanations and specific code references. Makes reviewing a breeze.

I tested with 9.6 and 9.5 and it works as described. Variable product appears from 9.6 👍 

>  find JN site with Beta Tester option + trunk in core the easiest way. The store should have at least one purchasable simple & variable product.

Plugin-wise, I personally started to enjoy Jetpack Beta Tester more than WooCommerce Beta, not sure which you use. In this case, Jetpack Beta Tester allowed me to switch between newest trunk, rc, and stable release on an existing site.

**Comment by jaclync** [comment]
> Thanks for an excellent description with explanations and specific code references. Makes reviewing a breeze.

The description was mostly from the GitHub Copilo Summary action, with some modifications after. The detailed references are pretty nice.

Screenshot 2024-12-12 at 8 11 54 AM


> Plugin-wise, I personally started to enjoy Jetpack Beta Tester more than WooCommerce Beta, not sure which you use. In this case, Jetpack Beta Tester allowed me to switch between newest trunk, rc, and stable release on an existing site.

Nice, I'll try that on my non-JN sites. Jirka had an issue with plugin version when his site had 2 installations of WC/WooPayments plugins from Jetpack Beta & standard installation p1732112938811029/1732014675.214429-slack-C025A8VV728. I think just making sure that there is only one plugin in the installed plugins should fix it.

**Comment by staskus** [comment]
> The description was mostly from the GitHub Copilo Summary action, with some modifications after. The detailed references are pretty nice.

Nice, I didn't know. I'll try it as well!

**Pull request merged**